### PR TITLE
SDKS-2239 - IG policy

### PR DESCRIPTION
**Pull request by jeyanthanperiyasamy**
# JIRA Ticket

[Please, link jira ticket here.](https://bugster.forgerock.org/jira/browse/SDKS-2239)

https://bugster.forgerock.org/jira/browse/SDKS-2239

# Description

- Create a new PolicyAdviceCreator class to address all the different routes
- Handle header response redirect from IG
- Update sample App to add the new header

# Definition of Done Checklist:

- [ ] Acceptance criteria is met.
- [ ] All tasks listed in the user story have been completed.
- [ ] Coded to standards.
- [ ] Ensure backward compatibility.
- [ ] API reference docs is updated.
- [ ] Unit tests are written.
- [ ] Integration tests are written.
- [ ] e2e tests are written.
- [ ] Functional spec is written/updated.
- [ ] Example code snippets have been added.
- [ ] Change log updated.
- [ ] Documentation story is created and tracked.
- [ ] Tech debts and remaining tasks are tracked in separated ticket(s).

### CODE REVIEW

File: FRAuth/FRAuth/Authorization/PolicyAdvice.swift [hunk 1]
```diff
@@ -128,15 +130,18 @@ public enum AdviceType: String {
...
             FRLog.w("Failed to parse AdviceType string value")
             return nil
         }
         
-        authIndexType = OpenAM.compositeAdvice
-        authIndexValue = "<Advices><AttributeValuePair><Attribute name=\"\(adviceType.rawValue)\"/><Value>\(value)</Value></AttributeValuePair></Advices>"
+        self.authIndexType = authIndexType ?? OpenAM.compositeAdvice
+        self.authIndexValue = authIndexValue ?? "<Advices><AttributeValuePair><Attribute name=\"\(adviceType.rawValue)\"/><Value>\(value)</Value></AttributeValuePair></Advices>"
         self.type = adviceType
         self.value = value
```
Comment (vahancouver) [inline FRAuth/FRAuth/Authorization/PolicyAdvice.swift:147]: Update copyright

File: FRAuth/FRAuth/Policy/AuthorizationPolicy.swift [hunk 2]
```diff
@@ -131,6 +131,13 @@ import Foundation
...
+        
+        if let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 401, let json = httpResponse.allHeaderFields["Www-Authenticate"] as? String,
+           let policyAdvice = PolicyAdviceCreator().parseAsBase64(advice: json) {
+            FRLog.i("[AuthorizationPolicy] PolicyAdvice JSON object found from response JSON payload; returning PolicyAdvice")
+            return policyAdvice
+        }
+        
         if let data = responseData, let json = try? JSONSerialization.jsonObject(with: data, options: []) as? [[String: Any]], let evalResult = json.first, let policyAdvice = PolicyAdvice(json: evalResult) {
             FRLog.i("[AuthorizationPolicy] PolicyAdvice JSON object found from response JSON payload; returning PolicyAdvice")
             return policyAdvice
```
Comment (vahancouver) [inline FRAuth/FRAuth/Policy/AuthorizationPolicy.swift:143]: Update copyright

File: FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/Authorization/PolicyAdviceTests.swift [hunk 3]
```diff
@@ -98,6 +110,20 @@ class PolicyAdviceTests: FRAuthBaseTest {
...
+        XCTAssertNotNil(advice)
+        XCTAssertNotNil(advice?.txId)
+        XCTAssertEqual(advice?.txId, "5afff42a-2715-40c8-98e7-919abc1b2dfc")
+        XCTAssertEqual(advice?.value, "5afff42a-2715-40c8-98e7-919abc1b2dfc")
+        XCTAssertEqual(advice?.type, AdviceType.transactionCondition)
+        XCTAssertEqual(advice?.authIndexType, "other_advice")
+        XCTAssertEqual(advice?.authIndexValue, indexValue)
         
         
         //  Given
```
Comment (vahancouver) [inline FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/Authorization/PolicyAdviceTests.swift:129]: Update copyright

File: FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/Policy/AuthorizationPolicyTests.swift [hunk 4]
```diff
@@ -285,6 +285,33 @@ class AuthorizationPolicyTests: FRAuthBaseTest {
...
+    
+    func test_18_redirect_evaluation_with_307_status_with_base64_encoded() {
+        let policy = AuthorizationPolicy(validatingURL: [URL(string: "https://openam.example.com/anything")!], delegate: self)
+        let header: [String: String] = ["Www-Authenticate": "SSOADVICE realm=\"/\",advices=\"eyJUcmFuc2FjdGlvbkNvbmRpdGlvbkFkdmljZSI6WyI1ODY2OWUxOS00MjVhLTQzMzMtOTFkOC03MDk5NWFmMDY5MjciXX0=\",am_uri=\"https://default.forgeops.petrov.ca/am/\""]
+        let response = HTTPURLResponse(url: URL(string: "https://openam.example.com/anything")!, statusCode: 401, httpVersion: nil, headerFields: header)!
+        let advice = policy.evaluateAuthorizationPolicy(responseData: nil, response: response, error: nil)
+        XCTAssertNotNil(advice)
+        XCTAssertEqual(self.list.count, 0)
+    }
 }
```
Comment (vahancouver) [inline FRAuth/FRAuthTests/FRAuthSwiftTests/FRAuth/Policy/AuthorizationPolicyTests.swift:317]: update copyright

**Review by vahancouver — CHANGES_REQUESTED** [review]
Copyright updates missing
overall changes look good

File: FRAuth/FRAuth/Authorization/PolicyAdviceCreator.swift [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+//  Copyright (c) 2023 ForgeRock. All rights reserved.
+//
+//  This software may be modified and distributed under the terms
+//  of the MIT license. See the LICENSE file for details.
+//
+
+
+import Foundation
+
+class PolicyAdviceCreator {
```
Comment (rodrigoareis) [inline FRAuth/FRAuth/Authorization/PolicyAdviceCreator.swift:14]: missing doc?

File: SampleApps/FRExample/FRExample/ViewController.swift [hunk 6]
```diff
@@ -162,7 +162,7 @@ class ViewController: UIViewController {
         FRURLProtocol.tokenManagementPolicy = policy
         
         //  - MARK: Authorization Policy - Example
-        let authPolicy = AuthorizationPolicy(validatingURL: [URL(string: "http://localhost:9888/policy/transfer")!], delegate: self)
+        let authPolicy = AuthorizationPolicy(validatingURL: [URL(string: "https://openig.petrov.ca/products")!], delegate: self)
```
Comment (rodrigoareis) [inline SampleApps/FRExample/FRExample/ViewController.swift:165]: should we revert this URL change?

**Review by rodrigoareis — CHANGES_REQUESTED** [review]
Changes looks good! Minor comments left

File: SampleApps/FRExample/FRExample/ViewController.swift (hunk 6 above)
Comment (jeyanthanperiyasamy) [reply SampleApps/FRExample/FRExample/ViewController.swift:165]: yea, thats the production one. Good call

**Review by rodrigoareis — APPROVED** [review]
Changes looks good to me

**Pull request merged**